File: experiment-14/runs/language=elixir_model=opus-4.8-fast_prompt=TDD/rep1/test/brazilian_soccer/queries/stats_test.exs
defmodule BrazilianSoccer.Queries.StatsTest do
  use ExUnit.Case, async: true

  alias BrazilianSoccer.Queries.Stats

  setup do
    {:ok, dataset: BrazilianSoccer.Fixtures.dataset()}
  end

  describe "summary/2" do
    test "computes averages and outcome rates", %{dataset: ds} do
      s = Stats.summary(ds)

      assert s.matches == 6
      assert_in_delta s.avg_goals_per_match, 2.0, 0.001
      assert_in_delta s.home_win_rate, 4 / 6, 0.001
      assert_in_delta s.draw_rate, 2 / 6, 0.001
      assert_in_delta s.away_win_rate, 0.0, 0.001
    end

    test "can be scoped to a competition", %{dataset: ds} do
      s = Stats.summary(ds, competition: "Copa do Brasil")
      assert s.matches == 1
    end
  end

  describe "biggest_wins/2" do
    test "ranks matches by goal margin", %{dataset: ds} do
      [top | _] = Stats.biggest_wins(ds, limit: 1)
      assert top.home_team == "Palmeiras"
      assert top.home_goals - top.away_goals == 2
    end
  end

  describe "best_record/3" do
    test "ranks teams by win rate with a minimum-match threshold", %{dataset: ds} do
      [best | _] = Stats.best_record(ds, :home, min_matches: 2)
      assert best.team == "Flamengo"
      assert best.played == 4
    end
  end
end

File: experiments/adrianco/experiment-5/runs/language=typescript_model=claude-opus-4-7_tooling=beads/rep2/tests/stats.test.ts
// Feature: Aggregate statistical analysis
import { describe, it, expect, beforeAll } from "vitest";
import { DataStore } from "../src/dataStore.js";
import { aggregateStats, biggestWins, topScoringTeams, bestRecord } from "../src/queries/stats.js";

let ds: DataStore;
beforeAll(() => {
  ds = new DataStore().load();
});

describe("Feature: Aggregate statistics", () => {
  describe("Scenario: average goals and home win rate", () => {
    it("Given Brasileirão matches, When computing aggregate stats, Then averages are in plausible football ranges", () => {
      const s = aggregateStats(ds.matches, { competition: "Brasileirão" });
      expect(s.matches).toBeGreaterThan(0);
      expect(s.averageGoalsPerMatch).toBeGreaterThan(1);
      expect(s.averageGoalsPerMatch).toBeLessThan(5);
      expect(s.homeWinRate).toBeGreaterThan(0.3);
      expect(s.homeWinRate).toBeLessThan(0.7);
      expect(s.homeWinRate + s.awayWinRate + s.drawRate).toBeCloseTo(1, 5);
    });
  });

  describe("Scenario: biggest wins are sorted by margin", () => {
    it("Given matches with results, When asking biggest wins, Then margins are non-increasing", () => {
      const wins = biggestWins(ds.matches, { limit: 5 });
      expect(wins.length).toBe(5);
      let prev = Infinity;
      for (const m of wins) {
        const margin = Math.abs((m.homeGoal ?? 0) - (m.awayGoal ?? 0));
        expect(margin).toBeLessThanOrEqual(prev);
        prev = margin;
      }
    });
  });

  describe("Scenario: top scoring teams", () => {
    it("Given the data, When asking top scoring teams in Brasileirão 2019, Then top team has > 50 goals", () => {
      const rows = topScoringTeams(ds.matches, { competition: "Brasileirão", season: 2019, limit: 5 });
      expect(rows.length).toBe(5);
      expect(rows[0].goalsFor).toBeGreaterThan(50);
    });
  });

  describe("Scenario: best home record", () => {
    it("Given Brasileirão 2019, When asking for best home record, Then results are sorted by win rate", () => {
      const rows = bestRecord(ds.matches, "home", {
        competition: "Brasileirão", season: 2019, minMatches: 10, limit: 5,
      });
      expect(rows.length).toBe(5);
      for (let i = 1; i < rows.length; i++) {
        expect(rows[i].winRate).toBeLessThanOrEqual(rows[i - 1].winRate);
      }
    });
  });
});
